src/monattack.cpp

```diff
@@ -1329,7 +1329,19 @@ static body_part body_part_hit_by_plant()
 bool mattack::growplants( monster *z )
 {
     for( const auto &p : g->m.points_in_radius( z->pos(), 3 ) ) {
-        // TODO: Make this sensible - it can destroy EVERYTHING
+
+        // Only certain kinds of terrain can be destroyed by this attack.
+        // Stairs, lava, card readers, etc. are unaffected.
+        for( auto flag : {
+                 "DIGGABLE", "ROAD", "RAIL", "TREE", "WALL"
+             } ) {
+            if( g->m.has_flag_ter( flag, p ) ) {
+                goto okay;
```

Comment:
You can put this loop into a separate function and return from within the loop so you avoid using goto.

Reply:
Why?